File: 047_DiameterOfBinaryTree.cpp
// Topic: Diameter of Binary Tree
// Description: Compute longest path (number of edges) between any two nodes.
// Approach: DFS returning height while tracking max diameter as left_height + right_height.

#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int bestDiameter;
int height(TreeNode* node)
{
    if (!node)
        return 0;
    int lh = height(node->left);
    int rh = height(node->right);
    bestDiameter = max(bestDiameter, lh + rh);
    return 1 + max(lh, rh);
}

TreeNode* buildLevel(const vector<string>& vals)
{
    if (vals.empty() || vals[0] == "null")
        return nullptr;
    TreeNode* root = new TreeNode(stoi(vals[0]));
    queue<TreeNode*> q;
    q.push(root);
    int i = 1;
    while (!q.empty() && i < (int)vals.size())
    {
        TreeNode* node = q.front();
        q.pop();
        if (i < (int)vals.size() && vals[i] != "null")
        {
            node->left = new TreeNode(stoi(vals[i]));
            q.push(node->left);
        }
        ++i;
        if (i < (int)vals.size() && vals[i] != "null")
        {
            node->right = new TreeNode(stoi(vals[i]));
            q.push(node->right);
        }
        ++i;
    }
    return root;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<string> vals1 = {"1", "2", "3", "4", "5"};
    vector<string> vals2 = {"3", "9", "20", "null", "null", "15", "7"};
    TreeNode* r1 = buildLevel(vals1);
    TreeNode* r2 = buildLevel(vals2);
    bestDiameter = 0;
    height(r1);
    cout << "Test 1 -> " << bestDiameter << "\n";
    bestDiameter = 0;
    height(r2);
    cout << "Test 2 -> " << bestDiameter << "\n";
}

/*
Time Complexity: O(n)
Space Complexity: O(h)
*/

File: 047_DiameterOfBinaryTree.py
# Topic: Diameter of Binary Tree
# Description: Compute longest path (number of edges) between any two nodes.
# Approach: DFS returning height while tracking max diameter as left_height + right_height.

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.diameter = 0

        def height(node):
            if not node:
                return 0
            lh = height(node.left)
            rh = height(node.right)
            self.diameter = max(self.diameter, lh + rh)
            return 1 + max(lh, rh)

        height(root)
        return self.diameter


if __name__ == "__main__":
    # Test 1
    root = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))
    print(f"Test 1 -> {Solution().diameterOfBinaryTree(root)}")

    # Test 2
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    print(f"Test 2 -> {Solution().diameterOfBinaryTree(root)}")

"""
Time Complexity: O(n)
Space Complexity: O(h)
"""
